src/java.base/share/classes/java/io/BufferedInputStream.java

```diff
@@ -643,9 +643,14 @@ private long implTransferTo(OutputStream out) throws IOException {
         if (getClass() == BufferedInputStream.class && markpos == -1) {
             int avail = count - pos;
             if (avail > 0) {
-                // Prevent poisoning and leaking of buf
-                byte[] buffer = Arrays.copyOfRange(getBufIfOpen(), pos, count);
-                out.write(buffer);
+                // trust all OutputStreams from java.io
+                if (out.getClass().getPackageName() == BufferedInputStream.class.getPackageName()) {
+                    out.write(getBufIfOpen(), pos, count);
+                } else {
+                    // Prevent poisoning and leaking of buf
+                    byte[] buffer = Arrays.copyOfRange(getBufIfOpen(), pos, count);
+                    out.write(buffer);
+                }
```

Comment:
```suggestion
                out.write(getBufIfOpen(), pos, count);
```

What do you think of passing the buffer as is?
`ByteArrayInputStream` passes the buffer without extra copies anyway: https://github.com/openjdk/jdk/blob/9a6ca233c7e91ffa2ce9451568b3be88ccd04504/src/java.base/share/classes/java/io/ByteArrayInputStream.java#L207-L213

Reply:
I think we need to hear @mkarg's opinion on this

Reply:
> What do you think of passing the buffer as is?

No, it should only do for trusted targets. BAIS has an issue in that area that should be fixed.

Reply:
The buffer in question is protected, so any subclass can directly access it. In other words, untrusted code can easily acoess the buffer, and it does not sound fair to add extra overhead to the method which was created for the performance reasons.

Does copyOfRange do any good here? Do you mean JDK should copy every buffer it passes to non-JDK code?

Reply:
> The buffer in question is protected, so any subclass can directly access it. In other words, untrusted code can easily acoess the buffer, and it does not sound fair to add extra overhead to the method which was created for the performance reasons.

If something extends BIS then it has access to the protected fields, this isn't a concern here. Instead, the concern is about passing a reference to the target output stream.

Reply:
@vlsi Yes, unless the JRE comes up with read-only buffers all untrusted code should get copies of JRE-internal buffers only to provide buffer poisoning and spying data located beyond range limits. Subclasses are free to do what they want with the inherited buffer (it is *their* buffer implicitly), but target output stream might be an injected bad guy that we must not trust in any regard.